caffeine@patapon.info/extension.js

```diff
@@ -332,53 +774,188 @@ class Caffeine extends PanelMenu.Button {
+            const type = window.get_window_type();
+            // Accept only normal window, ignore all other type (dialog, menu,...)
+            if(type === 0) {
+                // Add 100 ms delay to handle window detection
+                setTimeout(this._toggleWorkspace.bind(this), 100);
+            }
         });
+
+        // Check and toggle Caffeine
+        this._toggleWorkspace();
     }
 
-    _appWindowsChanged(app) {
+    _appWindowFocusChanged() {
+        let winTrack = Shell.WindowTracker.get_default();
+        let appId = null;
+        let app = winTrack.focus_app;
+
+        if(app)
+            appId = app.get_id();
+        if(this._appConfigs.includes(appId) && !this._isToggleInhibited(appId)){
+            this._manageScreenBlankState(true); // Allow blank screen
+            this._manageNightLight(false, true);
+            this.addInhibit(appId); // Inhibit app            
+        } else if (!this._appConfigs.includes(appId) && this._appInhibitedData.size !== 0){
+            this._manageScreenBlankState(true); // Allow blank screen
+            this._manageNightLight(true, true);
+            // Uninhibit all apps
+            this._appInhibitedData.forEach((data, id) => {
+                this.removeInhibit(id);
+            });
+        }
+    }
+
+    _appStateChanged(appSys, app) {
         let appId = app.get_id();
         let appState = app.get_state();
-        if (appState !== Shell.AppState.STOPPED) {
-            this.addInhibit(appId);
-            if (this._settings.get_enum(NIGHT_LIGHT_KEY) > ControlNightLight.NEVER && this._proxy.NightLightActive) {
-                this._proxy.DisabledUntilTomorrow = true;
-                this._night_light = true;
-            } else {
-                this._night_light = false;
-            }
-        } else {
-            this.removeInhibit(appId);
-            if (this._settings.get_enum(NIGHT_LIGHT_KEY) > ControlNightLight.NEVER && this._proxy.NightLightActive) {
-                this._proxy.DisabledUntilTomorrow = false;
-                this._night_light = true;
-            } else {
-                this._night_light = false;
+      
+        if(this._appConfigs.includes(appId)){            
+            // Block App state signal
+            appSys.block_signal_handler(this._appStateChangedSignalId);
+            
+            // Allow blank screen
+            this._manageScreenBlankState(true);
+
+            if (appState === Shell.AppState.STOPPED && this._isToggleInhibited(appId)){
+                this._manageNightLight(true, true);
+                this.removeInhibit(appId); // Uninhibit app
+            } else if (appState !== Shell.AppState.STOPPED && !this._isToggleInhibited(appId)) {
+                this._manageNightLight(false, true);
+                this.addInhibit(appId); // Inhibit app
             }
+            
+            // Add 200 ms delay before unblock state signal
```

Comment:
Can you explain why you need to add a timer here ?

Reply:
This is a workaround to solve an issue with unwanted multiple signals when an app start.

You were using "windows-changed" to detect windows event that call "_appWindowsChanged" function:
```
 windowsChangedId: app.connect('windows-changed',
                    this._appWindowsChanged.bind(this)),
```
([https://gjs-docs.gnome.org/shell01~0.1/shell.app#signal-windows-changed](https://gjs-docs.gnome.org/shell01~0.1/shell.app#signal-windows-changed)) 

Now I use a different method using "app-state-changed":
```
this._appStateChangedSignalId =
	this._appSystem.connect('app-state-changed', this._appStateChanged.bind(this)); 
```
([https://gjs-docs.gnome.org/shell01~0.1/shell.appsystem#signal-app-state-changed](https://gjs-docs.gnome.org/shell01~0.1/shell.appsystem#signal-app-state-changed))
    
In both case, for some reasons I got multiple events for few millisecond after the app started. 

For example, with "Black Box" app when starting (from logs):
```
App: com.raggesilver.BlackBox.desktop, state change: 1 (STARTING)
App: com.raggesilver.BlackBox.desktop, state change: 2 (RUNNING)
App: com.raggesilver.BlackBox.desktop, state change: 0 (STOPPED)
App: com.raggesilver.BlackBox.desktop, state change: 2 (RUNNING)
App: com.raggesilver.BlackBox.desktop, state change: 0 (STOPPED)
...
```
(test with `log("App: " + appId + ", state change: " + appState);`  in a dedicated testing function called by "windows-changed" or "app-state-changed")

With the timer, I choose to ignore for 200 ms the next following events, then I switch on again listener. So this way I keep only the first valid one.
Back in November, I got 10 signals at start for the same event mixing RUNNING, STOPPED and STARTING state. Sometimes it's creating multiple inhibited flags and the extension can get stuck so I introduce this solution (see original commit: [https://github.com/pakaoraki/gnome-shell-extension-caffeine/commit/bab83b0f1d578bd821ef1a6cd4f8864a9df909f6](https://github.com/pakaoraki/gnome-shell-extension-caffeine/commit/bab83b0f1d578bd821ef1a6cd4f8864a9df909f6) and [https://github.com/pakaoraki/gnome-shell-extension-caffeine/commit/76b3fc094603f4189eb31e49e4b94319f19b061a](https://github.com/pakaoraki/gnome-shell-extension-caffeine/commit/76b3fc094603f4189eb31e49e4b94319f19b061a)).

I'm not sure if it's a bug from the shell/gjs. I redo some test recently and it seems that's the situation is a bit better: I got less signals but still 4 consecutive RUNNING/STOOPED signals for "Black box" terminal app when starting.

Reply:
Thanks for the detailled answer! got it